helix-contract/contracts/ln/base/LnBridgeSourceV3.sol

```diff
@@ -0,0 +1,495 @@
+    // the configure information can be updated
+    struct TokenConfigure {
+        // pay to system for each tx
+        uint112 protocolFee;
+        // Used to penalise relayer for each slashed transaction
+        uint112 penalty;
+        uint8 sourceDecimals;
+        uint8 targetDecimals;
+    }
+    // registered token info
+    struct TokenInfo {
+        TokenConfigure config;
+        // zero index is invalid
+        // use this index to indict the token info to save gas
+        uint32 index;
+        address sourceToken;
+        address targetToken;
+        // accumulated system revenues
+        uint256 protocolFeeIncome;
+    }
+    struct TransferParams {
+        uint256 remoteChainId;
+        address provider;
+        address sourceToken;
+        address targetToken;
+        uint112 totalFee;
+        uint112 amount;
+        address receiver;
+        uint256 nonce;
+    }
+    // hash(remoteChainId, sourceToken, targetToken) => TokenInfo
+    mapping(bytes32=>TokenInfo) public tokenInfos;
+    // the token index is used to be stored in lockInfo to save gas
+    mapping(uint32=>bytes32) public tokenIndexer;
+    // amountWithFeeAndPenalty = transferAmount + providerFee + penalty < type(uint112).max
+    // the status only has the following 4 values
+    // status == 0: lockInfo not exist -> can update to status 1
+    // status == 1: lockInfo confirmed on source chain(has not been withdrawn or slashed) -> can update to status 2 or 3
+    // status == 2: lockInfo has been withdrawn -> can't update anymore
+    // status == 3: lockInfo has been slashed -> can't update anymore
+    // we don't clean lockInfo after withdraw or slash to avoid the hash collision(generate the same transferId)
+    // when we wan't to get tokenInfo from lockInfo, we should get the key(bytes32) from tokenIndex, then get tokenInfo from key
+    struct LockInfo {
+        uint112 amountWithFeeAndPenalty;
+        uint64 timestamp;
+        uint32 tokenIndex;
+        uint8 status;
+    }
+    // transferId => LockInfo
+    mapping(bytes32 => LockInfo) public lockInfos;
+
+    struct SourceProviderInfo {
+        uint112 baseFee;
+        uint16 liquidityFeeRate;
+        uint112 transferLimit;
+        bool pause;
+    }
+
+    // hash(remoteChainId, provider, sourceToken, targetToken) => SourceProviderInfo
+    mapping(bytes32=>SourceProviderInfo) public srcProviders;
+    // for a special source token, all the path start from this chain use the same panaltyReserve
+    // 1. when a lock tx sent, the penaltyReserves decrease and the penalty move to lockInfo.amountWithFeeAndPenalty
+    // 2. when withdraw liquidity, it tries to move this penalty lockInfo.amountWithFeeAndPenalty back to penaltyReserves
+    // 3. when the penaltyReserves is not enough to support one lock tx, the provider is paused to work
+    // hash(sourceToken, provider) => penalty reserve
+    mapping(bytes32=>uint256) public penaltyReserves;
```

Comment:
Since the penalty is shared from multi-chain, Indexer should calculate the penalty correctly.